feat: Add Hazelcast as an online store

mehmettokgoz opened the pull request:
**What this PR does / why we need it**:
This PR adds Hazelcast as an online store. The implementation cover the `online_write_batch`, `online_read`, `teardown` and `update` methods and adds interactive CLI to initialize Hazelcast template. 

Signed-off by: Mehmet Tokgöz[ mehmet.tokgoz@hazelcast.com](mailto:mehmet.tokgoz@hazelcast.com)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #
No related issue.

achals commented on sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/README.md:
```diff
@@ -0,0 +1,87 @@
+    ssl_cafile_path: /path/to/ca/file
+    ssl_certfile_path: /path/to/cert/file
+    ssl_keyfile_path: /path/to/key/file
+    ssl_password: <YOUR_SSL_PASSWORD>
```
nit, this should be read from the enrvironment instead of being hardcoded in the `feature_store.yaml` file. 

```suggestion
    ssl_password: ${SSL_PASSWORD} # This value will be read form the `SSL_PASSWORD` environment variable.
```

mehmettokgoz replied:
That's a good idea, thanks! Replaced the SSL password with environment variable at https://github.com/feast-dev/feast/pull/3523/commits/dad8fc414d91205c63c1048d956cafe9ae316be6.

achals commented on sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/README.md:
```diff
@@ -0,0 +1,87 @@
+    ssl_cafile_path: /path/to/ca/file
+    ssl_certfile_path: /path/to/cert/file
+    ssl_keyfile_path: /path/to/key/file
+    ssl_password: <YOUR_SSL_PASSWORD>
```
Same comment.

mehmettokgoz replied:
https://github.com/feast-dev/feast/pull/3523#discussion_r1144673527

achals commented on sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:
```diff
@@ -0,0 +1,317 @@
+
+        for entity_key, values, event_ts, created_ts in data:
+            entity_key_str = base64.b64encode(
+                serialize_entity_key(
```
For new online stores, we should just be using `entity_key_serialization_version=2`.

mehmettokgoz replied:
Replaced the config parameter with version 2 at https://github.com/feast-dev/feast/pull/3523/commits/dad8fc414d91205c63c1048d956cafe9ae316be6.

achals commented on sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:
```diff
@@ -0,0 +1,317 @@
+        entity_keys_str = {}
+        for entity_key in entity_keys:
+            entity_key_str = base64.b64encode(
+                serialize_entity_key(
```
same comment.

mehmettokgoz replied:
Replaced the config parameter with version 2 at https://github.com/feast-dev/feast/pull/3523/commits/dad8fc414d91205c63c1048d956cafe9ae316be6.

achals reviewed (commented):
Looks mostly good - going to do one more pass over when I have some bandwidth!

achals commented:
/ok-to-test

mehmettokgoz commented:
Hey @achals 👋 I've updated the PR and addressed your review comments. Let me know if you need my help! I've also tested manually using the template and the test workflow runs correctly.

achals reviewed (approved):
/ok-to-test
/lgtm
